### PR TITLE
Move `h2osno` limiter to ensure positive snow depth.

### PR DESCRIPTION
**Pull request by quantheory**
In `LakeHydrologyMod`, when applying the effects of sublimation and dew to snow, the `snow_depth` variable is set using the `h2osno` variable before that variable is limited to positive values. This commit changes the order of those statements to ensure that `snow_depth` is not accidentally set to a negative value.

It is important to note that in the existing code, only extremely small-magnitude negative values are typically produced, because of another limiter applied earlier in the code. Specifically, `h2osno` variable is set using

```
h2osno(c) = h2osno(c) - qflx_sub_snow(p)*dtime
```

and `qflx_sub_snow` is set using

```
qflx_sub_snow(p) = min(qflx_evap_soi(p), h2osno(c)/dtime)
```

Thus, `h2osno` can only be temporarily negative due to limited floating point precision, i.e. the fact that we can get a slightly negative output from

```
h2osno(c) - (h2osno(c)/dtime)*dtime
```

For this reason, this commit only produces roundoff-level changes in answers under typical conditions.

It is also known that invalid values of `snow_depth` and `snowdp` are implicated in some rare model crashes (see #5798). The causes are unknown, but in at least one case involved exponentially growing negative snow depth. It is possible that this change will prevent such crashes; investigation is ongoing at the time of this commit.

[non-BFB]

### CODE REVIEW

**Comment by quantheory** [comment]
Notifying @crterai @wlin7. Using print/assert statements, I have confidence that this change is needed to avoid small negative snow depth values being produced by this code. However, I'm still testing to see if this code is also responsible for the large negative values crashing the coupled run with the gustiness changes.

**Comment by quantheory** [comment]
In order to avoid dividing by a potentially nearly-zero number when setting snow depth, Xubin Zeng also suggests this change:

> replace "snow_depth(c) = snow_depth(c) * h2osno(c) / h2osno_temp" by
>                  "snow_depth(c) = h2osno(c) * min (snow_depth(c)/h2osno_temp, 1._r8/50._r8)"
>             This fix is physically based. the implication is that the snow density should be > 50 kg/m^3 or 1/density should be < 1/50 kg/m^3. 

I agree that this change should be made, but it should possibly be a separate PR, because the effect would be greater than roundoff level.

@mabrunke @xubinzeng

**Comment by quantheory** [comment]
Noting that @billsacks [made this change in CLM](https://github.com/ESCOMP/CTSM/commit/3b01f70c6775b2f22b4cae6153b79c8bf0bed9d1) just over four years ago. (According to the commit message, this was done as a preventative measure rather than due to any known problem in the model.)

**Comment by quantheory** [comment]
I have confirmed that not only is this section of code responsible for the negative values, but also can blow up an initially small negative value to a much larger negative value, which seems to be the ultimate cause of the crashes. Here's some output I generated from the time step and location (over Lake Erie) where the bad value occurs:

```
  28:  old snow_depth =  -4.336809803419874E-022
  28:  new snow_depth =  -3.746352330393036E+074
  28:  old h2osno =   9.175650230087428E-100
  28:  new h2osno =   7.926383719952902E-004
  28:  qflx_evap_soi =  -4.403546511084945E-007
  28:  qlfx_sub_snow =   0.000000000000000E+000
  28:  qlfx_dew_snow =   4.403546511084945E-007
  28:  do_capsnow =  F
  28:  ERROR: shr_assert_in_domain: snow_depth has invalid value 
  28:  -3.746352330393036E+074
```

This confirms that the the extreme negative value of $-3.7 \times 10^{74}$ is generated here, and that it arises due to division by the very small previous value of snowmass, $9.2 \times 10^{-100}$ (`h2osno_temp` is referred to as "old h2osno" in the above output).

@crterai @wlin7 @golaz This shows pretty definitively that this error is a preexisting error dating back to the original version of CLM that ELM was branched from. It is also not related to #5850. I could also use some guidance. The above output shows that the aforementioned fix from Xubin is needed as well, because even if we avoid negative `snow_depth` values, the same error can produce unrealistic positive values. That fix will cause changes that are larger than roundoff, but likely has negligible impact on climate. Should I add that fix to this PR or make a separate one?

**Comment by quantheory** [comment]
At @crterai's suggestion, I did check the snow depth values in the code in a short run with and without this change. Without this change, negative `snow_depth` always occurs within a few time steps, whereas with this change, there were no negative values over a 10 day run. So this PR does work for preventing those bad values.

**Review by crterai — APPROVED** [review]
Thanks for creating this quick PR @quantheory. It looks good to me. And as @quantheory notes in the comments, it is a bug fix that's already been made in clm as well.

**Comment by crterai** [comment]
Pinging @thorntonpe so that this PR is on your radar. Xubin Zeng also has a proposed fix that is in @quantheory's comment above, which we'd be interested to hear your take on.

**Review by bishtgautam — APPROVED** [review]

**Comment by wlin7** [comment]
Merged to next.

**Comment by rljacob** [comment]
@wlin7 it is to late to merge something to next for testing.

**Comment by wlin7** [comment]
> @wlin7 it is to late to merge something to next for testing.

Too bad. I thought cutoff time is 11 PM CT. Waiting for final approval for the gustiness PR. If next testing not picking up the new merges, we can push the merges to master the following day.

**Comment by rljacob** [comment]
The earliest suite starts at 10pm CT.

**Comment by rljacob** [comment]
removed from next with a force push.

**Comment by wlin7** [comment]
> removed from next with a force push.

How about leaving them on as today's merge to next? This and #5850. The Only other PR merged yesterday is #6054, tested by next today, and fixed SSP etc.

**Comment by wlin7** [comment]
> > removed from next with a force push.
> 
> How about leaving them on as today's merge to next? This and #5850. The Only other PR merged yesterday is #6054, tested by next today, and fixed SSP etc.

**Comment by wlin7** [comment]
Re-merged to next, and to master after the return of next testings.

**Pull request merged**
